Sources/PortalSwift/Core/Keychain/PortalKeychain.swift

```diff
@@ -0,0 +1,573 @@
+      metadata.namespaces[.solana] = solanaCurve
+    }
+
+    PortalKeychain.metadata = metadata
+
+    // Build the client's wallet metadata
+    var wallets: [PortalCurve: PortalKeychainClientMetadataWallet] = [:]
+    for wallet in client.wallets {
+      wallets[wallet.curve] = PortalKeychainClientMetadataWallet(
+        id: wallet.id,
+        curve: wallet.curve,
+        publicKey: wallet.publicKey,
+        backupShares: wallet.backupSharePairs.map { share in
+          PortalKeychainClientMetadataWalletBackupShare(
+            backupMethod: share.backupMethod,
+            createdAt: share.createdAt,
+            id: share.id,
+            status: share.status
+          )
+        },
+        signingShares: wallet.signingSharePairs.map { share in
+          PortalKeychainClientMetadataWalletShare(
+            createdAt: share.createdAt,
+            id: share.id,
+            status: share.status
+          )
+        }
+      )
+    }
+
+    // Build the client's metadata
+    let clientMetadata = PortalKeychainClientMetadata(
+      id: client.id,
+      addresses: [
+        .eip155: client.metadata.namespaces.eip155?.address,
+        .solana: client.metadata.namespaces.solana?.address,
+      ],
+      custodian: client.custodian,
+      wallets: wallets
+    )
+
+    // TODO: Remove this when we go fully async and chain agnostic
+    self.legacyAddress = client.metadata.namespaces.eip155?.address
+    self.clientId = client.id
+
+    // Write the metadata to the keychain
+    try await self.setMetadata(clientMetadata)
+  }
+
+  public func setMetadata(_ withData: PortalKeychainClientMetadata) async throws {
+    guard let client = try await client else {
+      self.logger.error("PortalKeychain.setMetadata() - Client not found")
+      throw KeychainError.clientNotFound
+    }
+    let clientId = client.id
+    let data = try encoder.encode(withData)
+    guard let value = String(data: data, encoding: .utf8) else {
+      self.logger.error("PortalKeychain.setMetadata() - Unable to encode keychain data")
+      throw KeychainError.unableToEncodeKeychainData
+    }
+
+    try self.keychain.updateItem("\(clientId).\(self.metadataKey)", value: value)
+  }
+
+  public func setShares(_ withData: [String: PortalMpcGeneratedShare]) async throws {
+    guard let client = try await client else {
+      self.logger.error("PortalKeychain.setShares() - Client not found")
+      throw KeychainError.clientNotFound
+    }
+    let clientId = client.id
+    let data = try encoder.encode(withData)
+    guard let value = String(data: data, encoding: .utf8) else {
+      self.logger.error("PortalKeychain.setShares() - Unable to encode keychain data.")
+      throw KeychainError.unableToEncodeKeychainData
+    }
+
+    try self.keychain.updateItem("\(clientId).\(self.sharesKey)", value: value)
+  }
+
+  /*******************************************
+   * Private functions
+   *******************************************/
```

Comment:
We're now using a `PortalKeychainAccess` class (so it can be mocked and injected) to read and write keychain data.